### PR TITLE
Linear rewrite of check()

### LINKED ISSUE #4: Linear rewrite of check()
Pick what you wish!

```diff
diff --git a/composer.json b/composer.json
index 3b57d85..770faea 100644
--- a/composer.json
+++ b/composer.json
@@ -14,7 +14,7 @@
         "composer-plugin-api": "^1.1.0"
     },
     "require-dev": {
-        "composer/composer": "^2.0.x-dev",
+        "composer/composer": "2.0.x-dev",
         "phpunit/phpunit": ">=4.8.35 <6.0",
         "codeclimate/php-test-reporter": "dev-master"
     },
diff --git a/src/UpdateHelper/UpdateHelper.php b/src/UpdateHelper/UpdateHelper.php
index 2ed2d9e..40d84e5 100644
--- a/src/UpdateHelper/UpdateHelper.php
+++ b/src/UpdateHelper/UpdateHelper.php
@@ -14,7 +14,7 @@ class UpdateHelper
 {
     /** @var Event */
     private $event;
-    /** @var IOInterface */
+    /** @var IOInterface|null */
     private $io;
     /** @var Composer */
     private $composer;
@@ -69,42 +69,39 @@ class UpdateHelper

     public static function check(Event $event)
     {
-        if ($event instanceof ScriptEvent || $event instanceof PackageEvent) {
-            $io = $event->getIO();
-            $composer = $event->getComposer();
-            $autoload = __DIR__.'/../../../../autoload.php';
-
-            if (file_exists($autoload)) {
-                include_once $autoload;
-            }
-
-            $classes = static::getUpdateHelperConfig($composer);
+        if (!($event instanceof ScriptEvent) && !($event instanceof PackageEvent)) {
+            return;
+        }

-            foreach ($classes as $file => $class) {
-                $error = null;
+        $io = $event->getIO();
+        $composer = $event->getComposer();
+        $autoload = $composer->getConfig()->get('vendor-dir').'/autoload.php';

-                if (is_string($class) && class_exists($class)) {
-                    try {
-                        $helper = new $class();
-                    } catch (\Exception $e) {
-                        $error = $e->getMessage()."\nFile: ".$e->getFile()."\nLine:".$e->getLine()."\n\n".$e->getTraceAsString();
-                    } catch (\Throwable $e) {
-                        $error = $e->getMessage()."\nFile: ".$e->getFile()."\nLine:".$e->getLine()."\n\n".$e->getTraceAsString();
-                    }
+        if (file_exists($autoload)) {
+            include_once $autoload;
+        }

-                    if (!$error && $helper instanceof UpdateHelperInterface) {
-                        $helper->check(new static($event, $io, $composer));
+        $classes = static::getUpdateHelperConfig($composer);

-                        continue;
-                    }
-                }
+        foreach ($classes as $file => $class) {
+            if (!class_exists($class)) {
+                $io->writeError(JsonFile::encode($class).' is not a class in file '.$file);
+                continue;
+            }

-                if (!$error) {
-                    $error = JsonFile::encode($class).' is not an instance of UpdateHelperInterface.';
-                }
+            try {
+                $helper = new $class();
+            } catch (\Exception | \Throwable $e) {
+                $io->writeError($e->getMessage()."\nFile: ".$e->getFile()."\nLine:".$e->getLine()."\n\n".$e->getTraceAsString());
+                continue;
+            }

-                $io->writeError('UpdateHelper error in '.$file.":\n".$error);
+            if (!($helper instanceof UpdateHelperInterface)) {
+                $io->writeError(JsonFile::encode($class).' is not an instance of UpdateHelperInterface in file '.$file);
+                continue;
             }
+
+            $helper->check(new static($event, $io, $composer));
         }
     }

@@ -157,7 +154,7 @@ class UpdateHelper
     }

     /**
-     * @return IOInterface
+     * @return IOInterface|null
      */
     public function getIo()
     {
```

### PR DESCRIPTION
**Pull request by szepeviktor**
Fixes #4

### CODE REVIEW

File: src/UpdateHelper/UpdateHelper.php [hunk 1]
```diff
@@ -69,42 +69,39 @@ protected static function getUpdateHelperConfig(Composer $composer, $key = null)
...
+                $io->writeError(JsonFile::encode($class).' is not a class in file '.$file);
+                continue;
+            }
 
-                if (!$error) {
-                    $error = JsonFile::encode($class).' is not an instance of UpdateHelperInterface.';
-                }
+            try {
+                $helper = new $class();
+            } catch (\Exception | \Throwable $e) {
```
Comment (kylekatarnls) [inline src/UpdateHelper/UpdateHelper.php]: This syntax appears in PHP 7.1
Update-Helper needs to support 5.3.0

**Review by kylekatarnls — CHANGES_REQUESTED** [review]
Thanks, it raised the fact that I never finished my unit tests on this library. So I will push them when I have some time. Yet I see a multiple exceptions types catch that PHP < 7.1 does not support: `\Exception | \Throwable`

File: src/UpdateHelper/UpdateHelper.php (hunk 1 above)
Comment (szepeviktor) [reply src/UpdateHelper/UpdateHelper.php]: Yes https://github.com/kylekatarnls/update-helper/blob/master/composer.json#L15

Comment (szepeviktor) [reply src/UpdateHelper/UpdateHelper.php]: Done.

**Review by kylekatarnls — APPROVED** [review]

**Comment by szepeviktor** [comment]
> class_exists() expects parameter 1 to be string, array given

I may have removed is_string :(

**Comment by szepeviktor** [comment]
How can be a class name an array?

**Comment by szepeviktor** [comment]
I think the changed error messages make unit tests [fail](https://github.com/szepeviktor/update-helper/blob/static-analysis/tests/UpdateHelper/UpdateHelperTest.php#L180). How to resolve it?

**Comment by kylekatarnls** [comment]
The unit test show we lost the catch with "UpdateHelper error in" wrapper.

The class name(s) config in composer.json can be 1 class or an array of classes.

File: src/UpdateHelper/UpdateHelper.php [hunk 2]
```diff
@@ -71,42 +71,42 @@ protected static function getUpdateHelperConfig(Composer $composer, $key = null)
...
+                $helper = new $class();
+            } catch (Exception $e) {
+                $io->writeError($e->getMessage()."\nFile: ".$e->getFile()."\nLine:".$e->getLine()."\n\n".$e->getTraceAsString());
+                continue;
+            } catch (Throwable $e) {
+                $io->writeError($e->getMessage()."\nFile: ".$e->getFile()."\nLine:".$e->getLine()."\n\n".$e->getTraceAsString());
+                continue;
+            }
 
-                $io->writeError('UpdateHelper error in '.$file.":\n".$error);
```
Comment (kylekatarnls) [inline src/UpdateHelper/UpdateHelper.php:108]: Here we lost the "UpdateHelper error", as different packages could fail in Composer, a prefix for errors is a good way to identify the package that throws it.

**Review by kylekatarnls — CHANGES_REQUESTED** [review]

File: src/UpdateHelper/UpdateHelper.php (hunk 2 above)
Comment (szepeviktor) [reply src/UpdateHelper/UpdateHelper.php:108]: I see.

Comment (szepeviktor) [reply src/UpdateHelper/UpdateHelper.php:108]: Could you make the test pass?

Comment (kylekatarnls) [reply src/UpdateHelper/UpdateHelper.php:108]: I found a way that remove the duplications and keep the error prefix.

**Review by kylekatarnls — APPROVED** [review]

**Comment by szepeviktor** [comment]
🚢

**Pull request merged**